123-job-jwt/proposal.md

```diff
@@ -0,0 +1,175 @@
+          test "((hvjwt:some_secret.username))" = "hello-im-a-username"
+          test "((hvjwt:"secret with_spaces.etc".some_key))" = "some-value"
+
+```
+
+This would represent a reasonable configuration. If it is determined that
+`auth_params` field in the `var_source.config.vault_config` schema should be
+re-used, then `auth_params: {role: "example", jwt: "((.CONCOURSE_CI_JWT))"}`
+might be a reasonable value, assuming that local environment variables can be
+referenced in such a way and we have pre-populated `CONCOURSE_CI_JWT` with the
+signed JWT during container pre-flight. However, this format would require more
+awareness on the part of the person configuring the pipeline, and the `jwt`
+field in the `auth_params` argument would always be the same, so it seems to
+make sense to abstract it in this case.
+
+Then, the call to the Vault API for the authentication step might look like the
+following:
+
+```sh
+#!/usr/bin/env bash
+auth_path=$(echo ${auth_path} | sed -E "s|^/*||g" | sed -E "s|/*$||g" |
+  sed -E "s|^auth/?||g" | sed -E "s|/?login$||g" | sed -E "s|/+|/|g"
+)
+resp=$(curl -s -X POST -H "X-Vault-Namespace: ${namespace}" \
+  --data @payload.json ${url}/v1/auth/${auth_path}/login
+)
+export VAULT_TOKEN=$(echo "${resp}" | jq -r ".auth.client_token")
+```
+
+or (forgive my golang pseudocode, rusty and untested):
+
+```golang
+// not dealing with errors or worrying about syntax, this is pseudocode
+// but please still make fun of my pseudocode, maybe I'll learn something
+func (ap *VarSources.Config.auth_path) FmtAuthPath() nil {
+  ap := strings.Trim(ap, "/")
+  ap := strings.TrimPrefix(ap, "auth/")
+  ap := strings.TrimSuffix(ap, "/login")
+  ap := path.Clean(ap)
+}
+type LoginData struct {
+  Role string `json:"role"`
+  JWT  string `json:"jwt"`
+}
+type LoginResp struct {
+  auth struct {
+    ClientToken string   `json:"client_token"`
+    Accessor    string   `json:"accessor"`
+    Policies    []string `json:"policies"`
+    Lease       int      `json:"lease_duration"`
+    Renewable   bool     `json:"renewable"`
+  } `json:"auth"`
+}
+func (c *VarSources.Config) VaultAuth() (token string, accessor string, err error) {
+  // I'm sure the vault package implements the auth method much better than this
+  client := &http.Client{}
+  uri := fmt.Sprintf("%s/v1/auth/%s/login", c.url, c.auth_path.FmtAuthPath())
+  jsonData := LoginData{
+    Role: c.auth_role,
+    JWT:  os.Getenv("CONCOURSE_CI_JWT")
+  }
+  var buf bytes.Buffer
+  err := json.NewEncoder(&buf).Encode(jsonData)
+  req, err := http.NewRequest("POST", uri, &buf)
+  req.Header.Add("X-Vault-Namespace", c.namespace)
+  resp, err := client.Do(req)
+  defer resp.Body.Close()
+  var dec LoginResp
+  dec = json.NewDecoder(&buf).Decode(resp)
+  return dec.auth.ClientToken, dec.auth.Accessor, nil
+}
+```
+
+## Open Questions
+
+- Where should the claims be aggregated and signing portion be completed? (atc?)
+- How should the signed JWT be delivered to the container runtime?
```

Comment:
Yeah, that's more the question. `atc` would be doing the generation, but then how does it send it to the `containerd` or `runc` environment. The JWT is itself a secret - and while we're making it unique to the context each time, making it short-lived and limited in scope, ideally there is a way to protect it in transit or ensure that it all happens in protected memory space.

I haven't looked deeply at the `atc` code yet - wanted to make sure I was going the right direction here first - but I assume there are other data being passed to the containers like resources and everything else. So we may just want to leverage the same methods, assuming it's protected by TLS or encrypted or something. I was also curious if it might be a `tsa` use case, or if it should be delivered via `worker/baggageclaim` or what. (Yes, I'm this new to the internals of Concourse.)